### PR TITLE
Unit conversion rules

### PR DESCRIPTION
**Pull request by MichaelClerx**
See #156

### CODE REVIEW

**Comment by MichaelClerx** [comment]
Hi @jonc125 !

Is this the kind of approach you had in mind?
I'm assuming we want all the numbers and variables as pint quantities, but not sure that plays well with sympy?

File: fc/parsing/actions.py [hunk 1]
```diff
@@ -1452,6 +1474,128 @@ def get(ref):
...
+
+            # Convert expression to sympy
+            try:
+                expr = expr.to_sympy(lambda x: variable_generator(x, free), number_generator)
+            except ProtocolError as e:
+                print('  _[]_')
+                print('W (")')
+                print('|-(:\')-<')
+                print('|(\'  )')
+                print('JON HOW DO I RAISE A WARNING? ' + e.short_message)
```
Comment (jonc125) [inline fc/parsing/actions.py:1574]: pycml used to just print to stderr. Using the logging module would be a nicer approach!

**Comment by jonc125** [comment]
Which come to think of it might mean we need a way to convert 
`simple_expr` to a Python expression rather than Sympy to do both the 
pint unit arithmetic and Sympy algebra together? Or maybe create a 
restricted version of that for conversion rules, since I can't see rules 
actually needing anything beyond plus/minus/times/divide/pow as operators?

**Comment by jonc125** [comment]
Thinking further, the `to_sympy` method for those operators could just use Python operations, then we could reuse that - the Python operations would result in Sympy expressions if applied to Sympy atoms/symbols.

**Comment by MichaelClerx** [comment]
> Not had a chance to properly look at this. But I was hoping we could make the conversion rule behave such as to provide a get_conversion_factor result that would have magnitude a sympy expression implementing the rule, with units containing the pint Unit version of the rule, or something like that?

Should we maybe work towards some new syntax where you just provide a factor that can be used both ways instead of two lambdas with the hidden assumption that they can be reduced to multiplication by some factor?

**Comment by jonc125** [comment]
> Should we maybe work towards some new syntax where you just provide a factor that can be used both ways instead of two lambdas with the hidden assumption that they can be reduced to multiplication by some factor?

All the examples in the pint docs match that assumption. I'm not sure if you'd never want to raise to a power or take a root though? Or maybe even convert to a log scale? But such things might be better modelled explicitly rather than using a context/conversion rule?

File: fc/parsing/actions.py (hunk 1 above)
Comment (MichaelClerx) [reply fc/parsing/actions.py]: OK someone can do that as part of #12 then

**Comment by MichaelClerx** [comment]
> I'm not entirely convinced by reducing all conversion rules to multiplicative factors (it wouldn't work for the pint spectroscopy example `[length] <-> [frequency]: speed_of_light / n / value` or other transformations requiring reciprocals) but it's reasonable for a first implementation I think since it covers our existing use cases.

As far as I can tell `cellmlmanip` currently only supports multiplicative factors

**Comment by MichaelClerx** [comment]
I don't like it either, because it's not consistent with the interface we present to the user, but agreed should be OK for now

**Pull request merged**